### ISSUE TITLE
Wishlist: A family of "assert" functions

### ISSUE THREAD

**Issue by d-torrance** [Macaulay2/M2#1736]
For helping to diagnose test failures from build logs, it would be nice if we could see *why* an `assert` statement threw an error.  Right now, `assert` just takes a boolean expression as input and throws an error if it's anything but `true`.  Instead, we could have a family of functions that know a little bit more about the input and could include that information in the error message, e.g., something like:

```m2
i1 : assertEqual(2 + 2, 5)
stdio:1:1:(3): error: assertion failed: 4 != 5
```

**Comment by DanGrayson** [comment]
Can't you just look at the source code to see the statement that was false?

**Comment by d-torrance** [comment]
> Can't you just look at the source code to see the statement that was false?

Sure, but that only tells part of the story story.  I got this idea looking at a PPA build failure on Ubuntu 21.04 arm64 from last night:

```m2
testing: timing-quotient.m2
timing-quotient.m2:223:1:(3):[4]: error: assertion failed
```

Here's the code:

```m2
standardSecond = first timing (value Core#"private dictionary"#"spin") 2833 -- about 1 second on a Macbook Pro in 2015

...

P=QQ[x,y,z,MonomialOrder=>Lex];
d=z^4+z^2*x*y^9+z*x^9*y+x^5*y^5;
phi=map(P,P,matrix{{x^13*y^4,x^3*y,x^20*y^6*z}});
tim = timing factor(phi(d));
assert(tim#0 < .05 * standardSecond)
```

From the source code alone, I have no idea what `tim` and `standardSecond` actually were for this build.

**Comment by DanGrayson** [comment]
Maybe a better way to accomplish the same objective would be to change `assert` to a method function that converts expressions to values using `value`.  Then the code could say something like `assert BinaryOperation {symbol < , 4 , 3}` and the error message could display the expression, not its value.  You'd have to make `value BinaryOperation {symbol < , 4 , 3}` work, too, but that would be good, anyway.

**Comment by mahrud** [comment]
@d-torrance sorry if this is too late, but another option is to have assert also accept a string, then feed it to `capture` for evaluation and check the result, then if it fails it can actually print the string! Something like this:
```m2
assert' = method()
assert' String := s -> if first capture concatenate("assert(", s, ")") then error s
```
This is what it would look like:
```m2
i3 : assert'("2 == 2")

i4 : assert'("2 != 2")
currentString:1:1:(3):[5]: error: assertion failed
stdio:6:77:(3):[1]: error: 2 != 2
```

**Comment by d-torrance** [comment]
> @d-torrance sorry if this is too late, but another option is to have assert also accept a string, then feed it to `capture` for evaluation and check the result, then if it fails it can actually print the string!

That would save the step of looking at the source code, but still wouldn't give us very much information.  For example:

```m2
i235 : foo = () -> 5;

i236 : assert' "foo() == 4"
currentString:1:1:(3):[5]: error: assertion failed
stdio:392:77:(3):[1]: error: foo() == 4
```

Just looking at the error message alone, we have no idea what the value of `foo()` is.  But using #1738, we have:

```m2
i8 : assertEqual(foo(), 4)
stdio:8:1:(3): error: assertion failed:
5==4 is false
```

**Comment by d-torrance** [comment]
This was addressed in #1738.  Closing.
